include/revng/Pipeline/LLVMContainer.h

```diff
@@ -0,0 +1,201 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <cstdint>
+#include <memory>
+#include <optional>
+#include <type_traits>
+
+#include "llvm/ADT/DenseSet.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Analysis/ScalarEvolutionExpressions.h"
+#include "llvm/Bitcode/BitcodeWriter.h"
+#include "llvm/IR/DebugInfoMetadata.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/IR/Instruction.h"
+#include "llvm/IR/Metadata.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/Value.h"
+#include "llvm/IR/ValueHandle.h"
+#include "llvm/IRReader/IRReader.h"
+#include "llvm/Linker/Linker.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Transforms/Utils/Cloning.h"
+
+#include "revng/Pipeline/Context.h"
+#include "revng/Pipeline/InspectableLLVMContainer.h"
+#include "revng/Pipeline/LLVMPipe.h"
+#include "revng/Support/Assert.h"
+#include "revng/Support/FunctionTags.h"
+
+namespace Pipeline {
+
+/// Base class that provides a reasonable
+/// default inplmentation for a llvm container
```

Comment:
> reasonable default

AFAIU this is our implementation and that's it. If you want to use something else you won't be able to use anything else in this file.

Why not push all of this in `LLVMContainerBase`?

Reply:
not sure what you are referring to exactly but everything built on top of llvm containers can accept the type of the llvm container as a template parameter. 

This is needed because if someones needs to operate on a llvm container that is not poluted by the containers (either from the current project, upstream or downstream), it can create a new llvm container that is just like the base one but with a differen type and thus different inspectors.

Reply:
For example? All the methods are final. I mean, there's not much to customize differently.

Reply:
as an example, if a user wishes to put a new kind on llvm functions that already have one, something like a kind to describe all the functions that come from a memory range in the original binary. if one uses the default llvm container, those functions will be tagged as isolated, and there is nothing he can do to avoid that. With this machinery one can introduce a new llvm container that behaves as the old one but it does not have the isolated function inspector, and he can introduce a new one that has ownership over the function in the rage.

Reply:
I've made another comment on how we could handle this, situation, let's continue the discussion there.